### ISSUE TITLE
[Question] rpn layer question

### ISSUE THREAD

**Issue by JieZou1** [yhenon/keras-frcnn#190]
Hi,

I noticed that when building rpn layer, only the number of filters are specified as num_anchors, as below:

    x_class = Conv2D(num_anchors, (1, 1), activation='sigmoid', kernel_initializer='uniform', name='rpn_out_class')(x)
    x_regr = Conv2D(num_anchors * 4, (1, 1), activation='linear', kernel_initializer='zero', name='rpn_out_regress')(x)

According to the paper, the regression output of rpn layer (the class label and boundingbox) should correspond to each anchor. In other words, one filter should correspond to one anchor.

But the above codes only make sure that there are k and 4*K outputs. How can we be sure that these k and 4*k outputs would form 1-to-1 correspondence to k anchors.

Many thanks and look forward to someone shedding the light on this issue.

Best Regards
Jie

**Comment by shinstra** [comment]
The naming is a bit deceptive here. num_anchors is the number of anchors with the same Centre point (k), not the total number of anchors (WHk). Since the operation is a convolution you end up with a tensor for x_class that is something like (W, H, k) and x_regr as (W, H, k*4). The total number of anchors add up - they are just spatially spread out. It’s a nice trick as the coordinate of and x_class/x_regr in a sense refers to the specific anchor it’s result is derived from.

**Comment by JieZou1** [comment]
@shinstra Many thanks for your explanation. I understand num_anchors is for each individual center point, and I understand that the output shape of rpn is (W, H, k) and (W, H, k*4).

What I confused is, even though the total number adds up, how could we be sure that each one of them is from each anchor (i.e, each scale and aspect ratio)? I understand that training data for rpn is generated according to the predefined anchors (scales and aspect ratios), would this guarantees that during testing, the region proposals (candidates) would be generated according to the same anchor layout?

**Comment by oak-tree** [comment]
Hey @JieZou1 
When generating `train data`, the `calc_rpn` function responsible of mapping each `rpn` with the right.
General speaking the `rpn` part of the  network `learns` the `offset` from the each `anchor`  and if there is an object or not. When `testing`  the `rpn_to_roi` is being called. It uses the learned `offeset` in order to produce the right `bounding box`

**Comment by JieZou1** [comment]
@oak-tree Thanks. I understand those 2 functions. My confusing was how we can be sure that the k (class_label) and 4k outputs (bbox) from an position would be from each of k anchors. Now, I think I got the point: since the training samples are created according to a predefined anchor layout, the regression outputs must also follow the predefined anchor layout. We are just doing k regression functions at once.

@shinstra and @oak-tree Many thanks again for helping me on figuring this out. It should be obvious, but I was just struck there when I was first thinking about this issue.

BTW, In the original paper, rpn predicts 2*k and 4*k outputs. do you know why this implementation does not prediction probability? My guess is the probability of candidates are not important, since the confidence would be estimated in the classification layer anyway? What do you guys think? Many thanks.

**Comment by oak-tree** [comment]
@JieZou1 np :) 

about your question. the `rpn` predicts  for `k` instead of `2k` because it uses binary classification what ever is there an object or not.  The original article do that `binary` task with two neurons instead of one

![image](https://user-images.githubusercontent.com/667153/31318000-8f78c3e6-ac53-11e7-9e1a-7aa48c89773c.png)

**Comment by JieZou1** [comment]
@oak-tree Okay, I see. Many thanks.
